pep/app/authenticators/credentials.ts

```diff
@@ -0,0 +1,164 @@
+
+    authenticationHeaders = {
+        'Content-Type': 'application/json'
+    };
+    private _invalidateTimeout?: EmberRunTimer;
+
+    /**
+     * Calculate expiration time - copied from ember simple auth
+     *
+     * @param {number} expiresIn
+     * @returns
+     * @memberof CredentialsAuthenticator
+     */
+    _absolutizeExpirationTime(expiresIn: number) {
+        if (!isEmpty(expiresIn)) {
+            return new Date(new Date().getTime() + expiresIn * 1000).getTime();
+        } else {
+            return undefined;
+        }
+    }
+
+    /**
+     * Calculate the expires at value and add it to the response so it gets saved in the cookie
+     *
+     * @param {PepSecureAuthenticatedData} response
+     * @returns
+     * @memberof CredentialsAuthenticator
+     */
+    setupExpiresAt(response: PepSecureAuthenticatedData) {
+        const expiresAt = this._absolutizeExpirationTime(response.SessionExpires);
+        if (expiresAt && !isEmpty(expiresAt)) {
+            Object.assign(response, { expiresAt: expiresAt });
+        }
+        return response;
+    }
+
+    /**
+     * Authenticates and logs the user in as well as schedules a session expiration based on the
+     * SessionExpires number that is sent back in the result
+     * @param  {String} username
+     * @param {String} password
+     */
+    async authenticate(username: string, password: string) {
+        try {
+            const sessionData = this.session.getUnauthenticatedSession();
+            const response = await this.ajax.request<PepSecureAuthenticatedData>(`${ENV.authBaseUrl}/Authenticate`, {
+                method: 'POST',
+                headers: this.authenticationHeaders,
+                body: this.ajax.stringifyData({
+                    grant_type: 'password',
+                    SessionId: sessionData?.SessionId,
+                    UserName: username,
+                    Password: password
+                })
+            });
+            if (response.IsValidLogon) {
+                this.session.clearUnauthenticatedSession();
+                response.SessionType = SessionType.CREDENTIALS;
+                const updatedResponse = this.setupExpiresAt(response);
+                this._scheduleSessionExpiration(updatedResponse);
+                return updatedResponse;
+            } else {
+                this.session.setUnauthenticatedSession(response);
+                return reject(response);
+            }
+        } catch (errors) {
+            return reject(errors);
+        }
+    }
+
+    /**
+     * Invalidates the local session and logs the user out
+     */
+    async invalidate(data: PepSecureAuthenticatedData) {
+        try {
+            const params = serializeQueryParams({ SessionId: data.SessionId });
+            await this.ajax.request(`${ENV.authBaseUrl}/Users/Logout?${params}`, {
+                method: 'POST',
+                headers: this.authenticationHeaders
+            });
+        } finally {
```

Comment:
not 100% sure on this, but i think in order to have this resolve instead of rejecting when the request fails you need to do:
```ts
try {
    //....
} catch(err) {
   return resolve();
} finally {
    this._cancelTimeout();
}
```
as `finally` blocks don't affect/control how/what the function returns, it just gives you a spot to put logic that should always happen before it returns. You could easily test this tho by just changing the logout endpoint to an invalid url so it fails.

Reply:
unless we want this to reject when the logout fails, but i thought there was a reason why we would want it to always resolve (i think due to the case where if your session needs to be invalidated after its already expired on the server, which would cause logout to fail, but we still need to successfully invalidate the local session that case.)